server/src/main/java/org/apache/iotdb/db/wal/buffer/WALBuffer.java

```diff
@@ -0,0 +1,426 @@
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * This buffer guarantees the concurrent safety and uses double buffers mechanism to accelerate
+ * writes and avoid waiting for buffer syncing to disk.
+ */
+public class WALBuffer extends AbstractWALBuffer {
+  /** Maximum number of WALEdits in one serialize task */
+  public static final int BATCH_SIZE_LIMIT = 100;
+
+  private static final Logger logger = LoggerFactory.getLogger(WALBuffer.class);
+  private static final IoTDBConfig config = IoTDBDescriptor.getInstance().getConfig();
+  private static final long SYNC_WAL_DELAY_IN_MS = config.getSyncWalDelayInMs();
+  private static final int WAL_BUFFER_SIZE = config.getWalBufferSize();
+  /** notify serializeThread to stop */
+  private static final WALEdit CLOSE_SIGNAL = new WALEdit(-1, new DeletePlan());
+
+  /** whether close method is called */
+  private volatile boolean isClosed = false;
+  /** WALEdits */
+  private final BlockingQueue<WALEdit> walEdits = new ArrayBlockingQueue<>(BATCH_SIZE_LIMIT * 10);
+  /** two buffers switch between three statuses (there is always 1 buffer working) */
+  // buffer in working status, only updated by serializeThread
+  private volatile ByteBuffer workingBuffer;
+  // buffer in idle status
+  private volatile ByteBuffer idleBuffer;
+  // buffer in syncing status, serializeThread makes sure no more writes to syncingBuffer
+  private volatile ByteBuffer syncingBuffer;
+  /** lock to provide synchronization for double buffers mechanism, protecting buffers status */
+  private final Lock buffersLock = new ReentrantLock();
+  /** condition to guarantee correctness of switching buffers */
+  private final Condition idleBufferReadyCondition = buffersLock.newCondition();
+  /** single thread to serialize WALEdit to workingBuffer */
+  private final ExecutorService serializeThread;
+  /** single thread to sync syncingBuffer to disk */
+  private final ExecutorService syncBufferThread;
+
+  public WALBuffer(String identifier, String logDirectory) throws FileNotFoundException {
+    super(identifier, logDirectory);
+    allocateBuffers();
+    serializeThread =
+        IoTDBThreadPoolFactory.newSingleThreadExecutor(
+            ThreadName.WAL_SERIALIZE.getName() + "(node-" + identifier + ")");
+    syncBufferThread =
+        IoTDBThreadPoolFactory.newSingleThreadExecutor(
+            ThreadName.WAL_SYNC.getName() + "(node-" + identifier + ")");
+    // start receiving serialize tasks
+    serializeThread.submit(new SerializeTask());
+  }
+
+  private void allocateBuffers() {
+    try {
+      workingBuffer = ByteBuffer.allocateDirect(WAL_BUFFER_SIZE / 2);
```

Comment:
Can the number of ByteBuffer.allocateDirect be controlled? If the number of ByteBuffer.allocateDirect is infinite, the memory will be used up?